### PR TITLE
Set DisplayTitle for subtitles

### PR DESCRIPTION
**Pull request by redSpoutnik**
This is my attempt to solve the jellyfin-web issue [157](https://github.com/jellyfin/jellyfin-web/issues/157).

**Changes**

When a *Title* metadata is available for a subtitle, set *DisplayTitle* to it and append other infos in square brackets:

E.g. **Sub Title [Eng] [Default] [Forced]**

If no *Title* metadata is available, then use previous behavior:

E.g. **Eng Default Forced**

**Issues**

[jellyfin-web#157](https://github.com/jellyfin/jellyfin-web/issues/157)

### CODE REVIEW

**Comment by nvllsvm** [comment]
Nice! Any ideas on making them both consistent? Both should have present the language, default, and forced attributes the same way.

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 1]
```diff
@@ -135,23 +135,49 @@ public string DisplayTitle
...
+                            attributes.Add("[" + StringHelper.FirstToUpper(Language) + "]");
+                        }
+                        else
+                        {
+                            attributes.Add("[Und]");
+                        }
+
+                        if (IsDefault)
+                        {
+                            attributes.Add("[Default]");
```
Comment (JustAMan) [inline MediaBrowser.Model/Entities/MediaStream.cs]: this should be localized

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 2]
```diff
@@ -135,23 +135,49 @@ public string DisplayTitle
...
-                        attributes.Add(StringHelper.FirstToUpper(Language));
+                        attributes.Add(Title);
+
+                        if (!string.IsNullOrEmpty(Language))
+                        {
+                            attributes.Add("[" + StringHelper.FirstToUpper(Language) + "]");
+                        }
+                        else
+                        {
+                            attributes.Add("[Und]");
```
Comment (JustAMan) [inline MediaBrowser.Model/Entities/MediaStream.cs]: this should be localized

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 3]
```diff
@@ -135,23 +135,49 @@ public string DisplayTitle
...
+                        }
+
+                        if (IsDefault)
+                        {
+                            attributes.Add("[Default]");
+                        }
+
+                        if (IsForced)
+                        {
+                            attributes.Add("[Forced]");
```
Comment (JustAMan) [inline MediaBrowser.Model/Entities/MediaStream.cs]: this should be localized

**Comment by EraYaN** [comment]
I feel the Kodi way would be neater:

`Language - Name - Codec - Channel Layout [flags]`
And when the Name is not set:
`Language - Codec - Channel Layout [flags]`
With flags being command separated values for default, forced etc.
So for example:
`Korean - DTS-HD MA 7.1 - 8 channel`
`Korean - Commentary - 2 channel - 2 channel`

For 2 channel the correct codec should be shown IMO, like LPCM/AAC/MP3/FLAC etc.

And the for subtitles the channel layout is not needed. And the type of subtitle is sufficient and we can add an 'internal' flag too.

We will also need to add all languages (probably should use the mkv language keys as keys.

I propose the format `<native> — <localized>` (that is an em-dash)
so for example
`English (US) — English (US)` in the en-US locale
or `Nederlands — Dutch` in the en-US locale
or `English (US) — Engels (V.S.)` in the nl-NL locale

This way you can always find your own language when the UI is in any other.

I can add all the languages to the json files and also all the stuff needed for the audio tracks. I suppose I can push to this branch.

**Comment by JustAMan** [comment]
An "ideal" solution for me would be sending to frontend the title of subtitles track and its properties (language code, enforced: yes/no, default: yes/no), and then frontend should work on displaying it with localization.

All IMHO.

**Comment by EraYaN** [comment]
Right that would also allow frond ends to go ham with icons/emoji (flags) etc.

We should probably define the correct Enums if they aren't there already instead of just passing strings.

**Comment by redSpoutnik** [comment]
Thank you all for your comments, I will try to do something better.
@JustAMan Should I store DisplayTitle tags (Und, Default, etc..) in the MediaStream itself or in a distinct object?

**Comment by JustAMan** [comment]
> Should I store DisplayTitle tags (Und, Default, etc..) in the MediaStream

Search me if I know (то есть хз 😄 ).
Maybe @EraYaN or @cvium could provide some input?

**Comment by EraYaN** [comment]
All types of mediastream (video, audio, sub) support these things in matroska files, so I think an "IsDefault" flag etc works.

**Comment by nvllsvm** [comment]
I'm all for a better approach, but I'd like to see this solution merged after the localization and consistency issues are fixed. It's low-risk and adds value without introducing the possibility of breakage in the future.

**Comment by redSpoutnik** [comment]
OK, for now I've added localization on server-side and few changes with subtitles tags:
Language, Default and Forced tags are now appended only if they are not already in Title, and separated with an hyphen.

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 4]
```diff
@@ -599,6 +602,9 @@ private MediaStream GetMediaStream(bool isAudio, MediaStreamInfo streamInfo, Med
             {
                 stream.Type = MediaStreamType.Subtitle;
                 stream.Codec = NormalizeSubtitleCodec(stream.Codec);
+                stream.localizedUndefined = _localization.GetLocalizedString("Undefined");
```
Comment (JustAMan) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs]: why have this in two places?

File: MediaBrowser.Model/Entities/MediaStream.cs [hunk 5]
```diff
@@ -220,6 +234,7 @@ private string GetResolutionText()
             return null;
         }
 
+        /*
         private string AddLanguageIfNeeded(string title)
```
Comment (JustAMan) [inline MediaBrowser.Model/Entities/MediaStream.cs]: can you remove dead code instead of commenting it? makes reading reviews harder...

**Comment by joshuaboniface** [comment]
There's also a couple merge conflicts that need to be fixed up if you can.

File: MediaBrowser.Model/Entities/MediaStream.cs (hunk 5 above)
Comment (redSpoutnik) [reply MediaBrowser.Model/Entities/MediaStream.cs]: Sure, sorry for that.

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs (hunk 4 above)
Comment (redSpoutnik) [reply MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs]: In fact, I am not sure how the ProbeResultNormalizer from MediaEncoder is used, but it creates subtitle MediaStreams, so I copied the behavior here by precaution. 
The SqliteItemRepository create MediaStreams used in itemdetails and videoosd screens.

**Comment by redSpoutnik** [comment]
@joshuaboniface Done :)

**Pull request merged**